### ISSUE TITLE
|2021 NIPS| Does Knowledge Distillation Really Work? 

### ISSUE THREAD

**Issue by gusdlf93** [gusdlf93/Paper_Survey#14]
논문 요약 :
"Teacher를 잘 모방하는 Student가 좋은 Student는 아닙니다. 하지만 대체로 Teacher를 잘 모방해야 Student가 Teacher와 유사한 성능을 보여줄 수는 있습니다. 이와 관련해서 다양하게 기존에 알려져있던 잘못된 통념이나 관념들을 바로 잡아준 논문입니다."

KD는 Teacher Model을 emulate하기 위해서, 소규모의 Student Model을 훈련하기 위해 널리 사용되는 기술입니다.
일반적으로는 KD가 학생의 Generalization 성능을 향상시킨다고 알려져 있지만, 동작 원리가 우리가 알던 것과는 괴리가 있으며, 심지어, Student가 충분히 커서 Teacher를 모방할 Capacity가 있음에도 불구하고, Teacher 모방에 실패하는 경우들이 많이 있음을 밝혔습니다.

최적화가 어려운 이유는 교사와 학생이 일치하지 못하는 것이 주요 원인입니다.
KD에서 사용하는 데이터셋의 detail은 Teacher와 Student가 밀접하게 일치하게 만드는데 있어서 어떠한 역할로서 작동하는지를 알 수 있습니다. 추가적으로, Teacher와 Student가 밀접하다는게, Student가 General한 모델을 배운다는 것 또한 아니라는 것을 밝힙니다.

Network는 Large + Deep 할수록  General한 representation을 배우기 좋습니다.
Network가 small + efficient하다면 보통은 적은 inductive biase 인해서, 훈련 데이터만으로는 General한 representation을 배우기는 어려우며, 훈련 데이터를 대표하는 특징정보들이나 그들을 모방하기에는 충분한 capacity를 가지고 있다고 볼 수 있습니다.

큰 앙상블 모델 (Teacher)는 작은 모델 (Student)로 그들의 지식을 이전할 수 있습니다.
이와 관련된 많은 논문들이 있지만, 반대로 아래 그림과 같이 Stuent가 Teacher를 충분히 모방할 수 있는 Capacity는 있음에도 불구하고, 모방에 실패하여 매우 다른 출력값을 갖게 될 수 있다는 연구 결과 또한 있습니다.

밑에 그림에서는 Teacher와 Student가 동일하게 Resnet-56으로 구성되었으며,
(a)실험에서는 Teacher와 Student가 동일하고, (b)에서는 Teacher가 앙상블 모델(3개), Student는 단일 모델(1개)를 사용해서 성능을 비교한 것입니다.
그러나 보시게 되면, Teacher가 큰 모델(Ensemble)일 때, Student가 Teacher를 fidelity(모방하려는 정도)가 높을수록 일반화 성능 또한 높아진다는 것을 알 수 있습니다.
이와는 별개로 Self-Distillation에서는 Student가 Teacher를 따라가려는 정도에 따라 성능의 변동이 있는 것을 알 수 있습니다
![image](https://user-images.githubusercontent.com/89114114/167069508-db5bcc75-d05f-46cf-b15a-7bf50e685213.png)

Self-Distillation에서는, Student가 Teacher의 모방에 실패해도 역설적으로 그 성능이 향상된다고 널리 알려져 있습니다.
그러나 fidelity의 향상에 따라 Generalization성능이 향상이 된다는 것을 위의 그래프를 통해 알 수 있습니다.
그리고, Student Network는 Teacher를 모방하려는 능력(fidelity)을 얻는것이 놀라울정도로 어렵다는 것을 알 수 있습니다.
이는 섹션 5에서 fidelity를 데이터셋의 보강을 통해서 해결할 수 있다는 가설을 세웁니다.
섹션 6에서는 낮은 fidelity가 최적화 문제로 인한 가설을 세우고 이와 관련된 조사를 진행합니다.
섹션 7에서는 KD가 종종 매우 제한된 지식을 Student에게 전달한다는 것을 밝힙니다.

**Knowledge Distillation Transfers Knowledge Poorly**
대규모 네트워크를 정제할 때 높은 fidelity를 갖는것이 불가능한 원인을 찾아보도록 하겠습니다.

가장 먼저, Teacher와 Student를 동일하게 LeNet-5로 설정합니다.
100 epoch 동안 MNIST + EMINST 훈련 세트로 Teacher를 훈련하여 84~86%의 성능을 얻습니다. Teacher와 Student는 99%이상 일치하며, 동일한 성능을 보여주게 됩니다.
![image](https://user-images.githubusercontent.com/89114114/167075073-7034556f-7adc-483a-bacc-70ebab270366.png)

이제, 더 복잡한 예시를 들어 보려고 합니다. Teacher와 Student를 동일하게 ResNet-56 모델로 설정합니다.
CIFAR-100 데이터셋에서 학습을 진행하는데, CIFAR-100은 증류 데이터를 보강하기 위한 데이터가 따로 없으므로, SN-GAN으로 CIFAR-100에 대해서 데이터를 추가적 해서 사용합니다.

데이터를 추가함에 따라서 fidelity는 증가하지만, 매우 미미하게 증가했을 뿐이며, 99%에는 한참 못 미칩니다.
ResNet-56은 LeNet-5보다 많은 매개변수를 가지고 있으므로, Student가 Teacher를 완벽하게 모방하기에는 충분히 많은 예제를 보지 못했을 수 있습니다. 즉 예제가 적어서 모방에 실패한게 아닌가 하는 가설에 대해서는 5.1섹션에서 자세히 다룹니다.

또한, MNIST실험과 마찬가지로 증류 데이터 세트가 증가함에 따라 학생의 정확도가 교사의 정확도에 가까워지는 것을 볼 수 있습니다.
MNIST실험과는 달리 Student의 정확도가 Teacher보다 높았는데, fidelity가 높아짐에 따라 성능이 유사해지는 것을 알 수 있습니다.
(Student가 의도치 않게 학습 방향을 잘 찾았다가 Teacher를 따라감에 따라 성능이 하락한다는 것입니다.)

**What can self-distillation tell us about knowledge distillation in general?**
Furlanello 논문에 따르면, KD에서 증류 절차를 실패하면, 학생이 교사의 성능을 뛰어 넘을 수 있음을 보였습니다.
즉, Student의 fidelity가 높다는 가정 하에서는 Student는 Teacher의 성능을 능가할 수 없습니다.
반면에, Ensemble Distillation에서 Teacher가 독립적으로 훈련된 학생보다 훨씬 Generalization 성능이 좋은 경우, Generalization 성능을 얻기 위해 Student는 Teacher를 모방하는 것이 바람직합니다.

또한, Teacher가 Generalization이 잘 되고, Student가 fidelity가 높아서 Teacher를 잘 모방한다 하더라도, Student가 Teacher의 Generalization 성능을 가져왔다는 것을 의미하는 것은 아닙니다.

**If distillation already improves generalization, why care about fidelity?**
KD가 Generalization 성능을 올리는데 도움이 되는 경우들이 있긴 하지만, fidelity는 여전히 중요합니다.
KD는 보통 강력한 모델을 소형 장치에 이식하기 위해서 제안되었습니다.
KD에서 generalization성능과 fidelity는 큰 연관이 있습니다.
Ensemble KD의 경우 Teacher모델과 Student모델 사이에는 상당한 성능 격차가 있는 경우가 많습니다.
Ensemble 수에 따라 Teacher와 Student간의 일반화 성능은 점점 벌어지게 됩니다.
Student의 fidelity를 높이는 것은 이러한 환경에서 Student와 Teacher간의 generalization 성능 차이를 줄이는 가장 확실한 방법입니다.
-> 근데 이게 말이 되는 것인가?
-> 서로 다른 Teacher들은 서로 다른 filter로부터 특징정보들을 추출하였기 때문에, fidelity를 높인다고 단순히 generalization이 올라갈 것이라는 마인드는 너무 가정이 naive한 것이 아닌가?
-> 데이터가 무시못할정도로 많아진다면 그게 가능해질 것인가?

KD는 해석 가능성과 신뢰성을 위해서도 사용이 되는데, 예를 들어서 의료 진단에 대한 통찰력을 제공하거나 텍스트에서 감정을 이해하기 위해, 매우 큰 모델에서 해석 가능한 작은 모델로 KD를 진행하기도 합니다.

**Possible causes of low distillation fidelity**

1.Student Capacity와 fidelity와의 연관성은 거의 없었습니다.
2.상기에 보고된 내용들은 Network Architecture에 크게 상관이 없이 전반적으로 나타나는 것으로 예상이 됩니다.
ResNet 기반의 architecture들 뿐만 아니라 VGG16에서도 관측이 됬기 때문이라고 합니다.
3.Mnist나 CIFAR 뿐만 아니라 ImageNet에서도 동일한 결과들을 관측할 수 있었습니다.
4.단순히 이미지 데이터 뿐만 아니라, 텍스트 데이터에서도 비슷한 현상을 관측했습니다.
5.Student가 Teacher의 training data에 대한 Representation을 학습한다고, Teacher의 Test data의 Representation 또한 학습할 수 있는 것이 아니라는 것을 확인했습니다.

**More distillation data lowers train agreement**
KD를 진행할 때, 광범위한 Data Augmetnation이나 GAN의 사용은 fidelity를 낮춘다는 것을 알 수 있습니다.

![image](https://user-images.githubusercontent.com/89114114/167087068-66a65f6b-68cc-4355-92e3-40e75100be8c.png)

또한, Batch Norm 대신에 Layer Norm을 사용해야 하는데, 그 이유는
Teacher Model과 Student Model이 정확히 동일한 weight를 가지고 있다고 가정해도, 누적된 통계의 차이로 인해서 Teacher와 Student가 서로 다른 예측을 진행할 수 있기 때문입니다.
또한, 학생이 Teacher의 Weight에서 시작하든 From scratch로부터 학습하든 큰 차이가 나지 않는 것을 발견하였습니다.

결론 :
fidelity가 높다고 accuracy가 높은 것은 아니고, fidelity만 잘 따라가면 되!라는 기존의 관념들은 너무 naive한 생각이었다는 것.
결국은 서로 다른 하이퍼 파라미터로부터 측정하게 되므로, 학습 데이터에 대해서 fidelity가 높다고 test 데이터에서 fidelity가 높다는 것을 의미하지도 않습니다.

student의 accuracy가 높다는게, distillation fidelity가 높다는 것을 의미하지는 않습니다.
distillation fidelity가 높다고 accuracy가 높지도 않습니다.
fidelity는 Ensemble을 증류할 때 calibration과 관계가 있습니다.
fidelity가 높은 학생이 가장 좋은 성능을 보여주는 것은 아니지만, 대체로 calibraiton 성능이 높게 나타납니다.
KD에서 최적화는 어려운 문제입니다. capacity와 관계 없이 잘 진행되지 않습니다.

**Comment by gusdlf93** [comment]
여기까지 읽어도 여전히 KD가 batch_size에 둔감한 원인이 증명되진 않은거 같습니다.
역시 loss surface를 하나하나 찍어봐야..

**Comment by gusdlf93** [comment]
[Knowledge Distillation.pdf](https://github.com/gusdlf93/Paper_Survey/files/8791484/Knowledge.Distillation.pdf)
